pkg/cvo/cvo.go

```diff
@@ -56,6 +57,61 @@ const (
 	maxRetries = 15
 )
 
+// FeatureGates contains flags that control CVO functionality gated by product feature gates. The
+// names do not correspond to product feature gates, the booleans here are "smaller" (product-level
+// gate will enable multiple CVO behaviors).
+type FeatureGates struct {
+	// UnknownVersion flag is set to true if CVO did not find a matching version in the FeatureGate
+	// status resource, meaning the current set of enabled and disabled feature gates is unknown for
+	// this version. This should be a temporary state (config-operator should eventually add the
+	// enabled/disabled flags for this version), so CVO should try to behave in a way that reflects
+	// a "good default": default-on flags are enabled, default-off flags are disabled. Where reasonable,
+	// It can also attempt to tolerate the existing state: if it finds evidence that a feature was
+	// enabled, it can continue to behave as if it was enabled and vice versa. This temporary state
+	// should be eventually resolved when the FeatureGate status resource is updated, which forces CVO
+	// to restart when the flags change.
+	UnknownVersion bool
+
+	// ResourceReconciliationIssuesCondition controls whether CVO maintains a Condition with
+	// ResourceReconciliationIssues type, containing a JSON that describes all "issues" that prevented
+	// or delayed CVO from reconciling individual resources in the cluster. This is a pseudo-API
+	// that the experimental work for "oc adm upgrade status" uses to report upgrade status, and
+	// should never be relied upon by any production code. We may want to eventually turn this into
+	// some kind of "real" API.
+	ResourceReconciliationIssuesCondition bool
+}
+
+func DefaultGatesWhenUnknown() FeatureGates {
+	return FeatureGates{
+		UnknownVersion: true,
+
+		ResourceReconciliationIssuesCondition: false,
+	}
+}
+
+func GetCvoGatesFrom(gate *configv1.FeatureGate) FeatureGates {
+	enabledGates := DefaultGatesWhenUnknown()
+	// This is analogical to VersionForOperatorFromEnv() from o/library-go but the import is pretty
+	// heavy for a single, simple os.Getenv wrapper, so we just inline the logic here.
+	operatorVersion := os.Getenv("OPERATOR_IMAGE_VERSION")
+	klog.Infof("Looking up feature gates for version %s", operatorVersion)
```

Comment:
Later commit removed this specific line because logging this would be excessive (we'd do it in a loop):

https://github.com/openshift/cluster-version-operator/blob/7bd309612f3f741a2479c7960299344de9f162bc/pkg/featuregates/featuregates.go#L79-L84

But one-time log of how feature gates were detected (what version do we see, if we found matching item in FeatureGate .status, etc) is definitely useful and we should have it. Right now we have a one-time log like this:

```console
$ /usr/bin/grep -e start.go -e featuregates.go -e stopper.go openshift-cluster-version_cluster-version-operator-77b9674d7-58k2s_cluster-version-operator.log
I0325 15:16:00.615836       1 start.go:23] ClusterVersionOperator v1.0.0-1183-gb6dbd748-dirty
I0325 15:16:00.747858       1 start.go:281] Waiting on 1 outstanding goroutines.
I0325 15:20:52.964166       1 start.go:549] FeatureGate found in cluster, using its feature set "" at startup
I0325 15:20:54.090127       1 start.go:574] CVO features for version 4.16.0-0.ci.test-2024-03-25-143851-ci-op-drnlyhbh-latest enabled at startup: {desiredVersion:4.16.0-0.ci.test-2024-03-25-143851-ci-op-drnlyhbh-latest unknownVersion:false resourceReconciliationIssuesCondition:false}
I0325 15:20:54.090240       1 featurechangestopper.go:123] Starting stop-on-features-change controller with startingRequiredFeatureSet="" startingCvoGates={desiredVersion:4.16.0-0.ci.test-2024-03-25-143851-ci-op-drnlyhbh-latest unknownVersion:false resourceReconciliationIssuesCondition:false}
```